### ISSUE TITLE
Updating dependencies doesn't flush cache

### ISSUE THREAD

**Issue by virresh** [coala/coala#5451]
### Overview Description
Whenever a linter is upgraded, the cache of coala is not flushed and new linting fixes introduced in the dependency are not shown

For e.g, take pycodestyle
```W605 invalid escape sequence '\s'``` was introduced in pycodestyle==2.4
So when coala is run with pycodestyle==2.3, the output is cached and thus doesn't show any results for W605

But even after updating to pycodestyle==2.4, the output is shown exactly same as before being cached

# Steps to Reproduce

Create a directory (```mkdir cacheTest```).
Make a file ```file.py``` having contents:
```
import re

reg = '#\s(\d+)'
string = '#123, # 123'

print(re.compile(reg).findall(string))
```

The contents of ```.coafile``` are:
```
[all]
files = file.py
bears = PycodestyleBear
```

1. Ensure ```pycodestyle``` is set to version 2.3
```
pip install pycodestyle==2.3
```
2. Run coala
```
coala
```
3. Now install ```pycodestyle``` version 2.4
```
pip install pycodestyle==2.4
```
4. Now again run coala
```
coala
```
5 Now run coala with flushed cache
```
coala --flush-cache
```

# Actual Results
The output is different after flushing the cache

# Expected Results
After updating to pycodestyle==2.4, there shouldn't have been a need to explicitly flush cache

# Reproducibility
100%

#### Additional Information:
A sample run on my machine:
```
(coala) viresh@viresh-PC:~/Desktop/coalaDebug$ pip install pycodestyle==2.3
Collecting pycodestyle==2.3
  Using cached https://files.pythonhosted.org/packages/32/09/8e580517e4af96141c41f531bca14c4408e7cb716864022737811aaac1be/pycodestyle-2.3.0-py2.py3-none-any.whl
Installing collected packages: pycodestyle
  Found existing installation: pycodestyle 2.4.0
    Uninstalling pycodestyle-2.4.0:
      Successfully uninstalled pycodestyle-2.4.0
Successfully installed pycodestyle-2.3.0
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


(coala) viresh@viresh-PC:~/Desktop/coalaDebug$ coala
Executing section all...
Executing section cli...


(coala) viresh@viresh-PC:~/Desktop/coalaDebug$ pip install pycodestyle==2.4
Collecting pycodestyle==2.4
  Using cached https://files.pythonhosted.org/packages/e5/c6/ce130213489969aa58610042dff1d908c25c731c9575af6935c2dfad03aa/pycodestyle-2.4.0-py2.py3-none-any.whl
Installing collected packages: pycodestyle
  Found existing installation: pycodestyle 2.3.0
    Uninstalling pycodestyle-2.3.0:
      Successfully uninstalled pycodestyle-2.3.0
Successfully installed pycodestyle-2.4.0
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


(coala) viresh@viresh-PC:~/Desktop/coalaDebug$ coala
Executing section all...
Executing section cli...


(coala) viresh@viresh-PC:~/Desktop/coalaDebug$ coala --flush-cache
Executing section all...

file.py
[   3] reg·=·'#\s(\d+)'
**** PycodestyleBear (W605) [Section: all | Severity: NORMAL] ****
!    ! W605 invalid escape sequence '\s'
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. Add (I)gnore comment
[    ] Enter number (Ctrl-D to exit): 0

file.py
[   3] reg·=·'#\s(\d+)'
**** PycodestyleBear (W605) [Section: all | Severity: NORMAL] ****
!    ! W605 invalid escape sequence '\d'
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. Add (I)gnore comment
[    ] Enter number (Ctrl-D to exit): 0
Executing section cli...
(coala) viresh@viresh-PC:~/Desktop/coalaDebug$ 
```

TL;DR :
Upgrading dependencies should flush the cache. It doesnt.
Cache needs to record something about the dependencies to trigger flushing.
Possibilities: Maybe executable file timestamp Or dep management needs to obtain version of installed dep and that should be stored in coala settings cache

difficulty/high
